lib/src/async_cache.dart

```diff
@@ -89,12 +89,13 @@ class AsyncCache<T> {
   }
 
   /// Removes any cached value.
-  void invalidate() {
+  Future invalidate() {
     _cachedValueFuture = null;
-    _cachedStreamSplitter?.close();
+    Future invalidate = _cachedStreamSplitter?.close();
```

Comment:
I'd just make the function async and await here – in theory you could be avoiding some unneeded async silly if the splitter != null – but you are also adding some complexity to how this method reads.

Maybe add an await and put the null check in a block – best of both worlds?

Reply:
Also this is a ~breaking change – folks will start getting warnings about awaiting here if they have the lint enabled.

Why is this change needed?

Reply:
Well, technically I would say this is a bugfix.

If the invalidate method does asynchronous work, and the future of that work is thrown away, its impossible to await invalidation. There is at least some code (in the tests) awaiting validation, and that silently did not actually await the invalidation.

However, this could be maintained as returning void, and we could add a `invalidateAsync` method which returns the future of `_cachedStreamSplitter?.close()` ?

Regarding the order of operations, I just really wanted to land a change that's as small as possible -- so this won't add asychrony in the void case (like before) and it won't change the order of operations vs before. Not necessarily important, and not as clean as `await`ing here, but I just didn't want spooky asynchrony changes to stall the CL.

If you aren't spooked by the await I'll do that though! And do you want it to be a new method? The good news is that dart 2 will not let them use the wrong method ("you can't await invalidate() as it returns void")

Reply:
ping @kevmoo

Reply:
I don't understand how this relates to an `await void`? Was something in the tests awaiting the result of this function before? Wasn't that effectively `await null`? Can it still be?

Reply:
Yeah, a test was awaiting this method. It *could* definitely return null, but I would consider the following a bug:

```
Future f() {
  spawn_async_operation_without_await();
  return null;
}
```

"function returning future doesn't actually block until work is complete" is the title I'd give that bug!

Alternatively, and I mean this as a strawman but I also honestly prefer it, I could do

```
Null f() {
  spawn_async_operation_without_await();
  return null;
}
```

Reply:
What I mean is, the old behavior was already translatable to `await null;` This isn't a case of a `Future` slipping through a `void` signature and being an otherwise valid `await`. This is a case of a `Null` going through a `void` signature. I agree on principle with the idea of making this async work _possible_ to await. But as it is this change is not maintaining the same semantics in the test.

Reply:
Yeah, I was careful to preserve the ordering semantics just out of conservatism, but didn't do so for this part, which is arguably inconsistent. I can leave a TODO and return null.